gateway/ln-gateway/src/ng/tests.rs

```diff
@@ -74,10 +78,83 @@ pub fn sha256(data: &[u8]) -> sha256::Hash {
 }
 
+async fn pay_valid_invoice(
+    invoice: Invoice,
+    user_client: &Client,
+    gateway: &Client,
+) -> anyhow::Result<()> {
+    // User client pays test invoice
+    let (pay_type, contract_id) = user_client.pay_bolt11_invoice(invoice.clone()).await?;
+    match pay_type {
+        PayType::Lightning(pay_op) => {
+            let mut pay_sub = user_client.subscribe_ln_pay(pay_op).await?.into_stream();
+            assert_eq!(pay_sub.ok().await?, LnPayState::Created);
+            let funded = pay_sub.ok().await?;
+            assert_matches!(funded, LnPayState::Funded);
+
+            let gw_pay_op = gateway.gateway_pay_bolt11_invoice(contract_id).await?;
+            let mut gw_pay_sub = gateway
+                .gateway_subscribe_ln_pay(gw_pay_op)
+                .await?
+                .into_stream();
+            assert_eq!(gw_pay_sub.ok().await?, GatewayExtPayStates::Created);
+            assert_matches!(gw_pay_sub.ok().await?, GatewayExtPayStates::Preimage { .. });
+            if let GatewayExtPayStates::Success {
+                preimage: _,
+                outpoint: gw_outpoint,
+            } = gw_pay_sub.ok().await?
+            {
+                gateway.receive_money(gw_outpoint).await?;
+            } else {
+                panic!("Gateway pay state machine was not successful");
+            }
+        }
+        _ => panic!("Expected Lightning payment!"),
+    }
+    Ok(())
+}
+
+#[tokio::test(flavor = "multi_thread")]
+async fn test_gateway_can_pay_ldk_node() -> anyhow::Result<()> {
+    // Running LDK Node with the mock services doesnt provide any additional
+    // coverage, since `FakeLightningTest` does not open any channels.
+    if !Fixtures::is_real_test() {
+        return Ok(());
+    }
+
+    gateway_test(|gateway, _, fed, user_client, bitcoin| async move {
+        let ldk = Fixtures::spawn_ldk(bitcoin.clone()).await;
+
+        ldk.open_channel(
+            Amount::from_msats(5000000),
+            gateway.node_pub_key,
+            gateway.listening_addr.clone(),
+            bitcoin.lock_exclusive().await,
+        )
+        .await?;
+
+        let gateway = gateway.remove_client(&fed).await;
+        // Print money for user_client
+        let (_, outpoint) = user_client.print_money(sats(1000)).await?;
+        user_client.receive_money(outpoint).await?;
+        assert_eq!(user_client.get_balance().await, sats(1000));
+
+        // Create test invoice
+        let invoice = ldk.invoice(sats(250), None).await?;
+        pay_valid_invoice(invoice, &user_client, &gateway).await?;
+
+        assert_eq!(user_client.get_balance().await, sats(1000 - 250));
+        assert_eq!(gateway.get_balance().await, sats(250));
+
+        Ok(())
+    })
+    .await
+}
+
 #[tokio::test(flavor = "multi_thread")]
 async fn test_gateway_client_pay_valid_invoice() -> anyhow::Result<()> {
     gateway_test(
-        |gateway, other_lightning_client, fed, user_client| async move {
```

Comment:
Seems to me like these would be better of in:

```rust
#[non_exhaustive]
Fixtures {
  gateway: Gateway,
  client: Client,
}
```

and then it could be:

```rust
gateway_test(|Fixtures { gateway, client, .. }| /* body */);
```

kind of thing.

Reply:
I agree, but could be a separate refactor imo.